### PR TITLE
Faster mapreduce for Broadcasted

### PR DESCRIPTION
**Pull request by tkf**
I suggest to add indexing-based specializations for `Broadcasted` in `mapreduce`.  As `Broadcasted` is almost like an array, it is mostly about "routing" `Broadcasted` to the right method.

I need to fill some more details, but it already works well for simple example:

```julia
julia> a = rand(10_000); b = rand(10_000);

julia> @btime sum(Broadcast.instantiate(Broadcast.broadcasted(*, $a, $b)));
  1.935 μs (2 allocations: 48 bytes)
```

which is somewhat comparable to `LinearAlgebra.dot`:

```julia
julia> @btime dot($a, $b);
  1.693 μs (0 allocations: 0 bytes)
```

Before it was 5x slower:

```julia
julia> @btime sum(Broadcast.instantiate(Broadcast.broadcasted(*, $a, $b)));
  10.067 μs (2 allocations: 48 bytes)
```

I'm opening a PR to get some feedback for polishing the implementation.  I'll ask specific questions below.

(Edit: now we require `Broadcast.instantiate` for fast reduce)

### CODE REVIEW

File: base/reduce.jl
Comment (tkf) [inline base/reduce.jl]: I needed to define this `AbstractBroadcasted` as broadcast.jl is loaded later than reduce.jl.  ATM it's just a hack to make this work, but I wonder if it makes sense to define

```julia
abstract type AbstractIndexable end
const Indexable = Union{AbstractArray, AbstractIndexable}
```

or even

```julia
abstract type Indexable{N} end
abstract type AbstractArray{T,N} <: Indexable{N} end
```

so that it's much easier for downstream projects to use indexing-based `mapreduce` implementation.

File: base/broadcast.jl
Comment (tkf) [inline base/broadcast.jl]: I need to look into how indexing is defined for `Broadcasted` (especially for how it interacts with `Axes`) and if this `IndexStyle` is correct.

However, it seems to me that it'd be much simpler if we define `IndexStyle(::Broadcasted)` than `IndexStyle(::Type{<:Broadcasted})` as it would let us use `eachindex(::Broadcasted)` in `IndexStyle`.  Does it makes sense to define `IndexStyle(::Broadcasted)` directly than the type trait?

File: base/reduce.jl
Comment (nalimilan) [reply base/reduce.jl]: Having an abstract type or a trait like `Indexable` would also be useful to support `mapreduce` over dimensions with `skipmissing` (https://github.com/JuliaLang/julia/pull/28027). It would avoid duplicating some methods.

Though a trait would probably be better than an abstract type, since it would be more flexible. In particular, an object returned by `skipmissing` could be marked as `Indexable` or not depending on whether it wraps an `Indexable` object (like an array) or another iterable. It would also allow objects that already inherit from a non-indexable abstract type to implement `Indexable` if they want, which wouldn't be possible with an abstract type.

File: base/broadcast.jl
Comment (nalimilan) [reply base/broadcast.jl]: AFAIK `IndexStyle` needs to take a type rather than an instance so that you can know statically what kind of indices to expect (that's more or less the definition of a trait). At least that's how its [documented](https://docs.julialang.org/en/latest/manual/interfaces/#man-interface-array-1).

Comment (tkf) [reply base/broadcast.jl]: My comment is based on the implementation detail in reduce.jl that `IndexStyle` is called on _instance_

https://github.com/JuliaLang/julia/blob/236c69b1a53acf4057de830979138765c79aa596/base/reduce.jl#L299

which is valid since `IndexStyle` is callable on `AbstractArray`:

https://github.com/JuliaLang/julia/blob/236c69b1a53acf4057de830979138765c79aa596/base/indices.jl#L70

However, to make `IndexStyle` a proper trait for `Broadcasted`, I defined methods for _types_ and added the method for instance on top of them, just like how it's done for `AbstractArray`:

https://github.com/JuliaLang/julia/blob/f4c68eeeeb14bfd03a0d53ea011ebe0e05b32037/base/broadcast.jl#L227

Note that direct definition of `IndexStyle(::Broadcasted)` would be type-stable if `eachindex` is so.  If `eachindex` is not type-stable we don't have the performance gain anyway.  Initially, I wasn't aiming for adding public interface for `Broadcasted` in this PR so I thought relying on the implementation detail that `IndexStyle` is only called on instance was OK.  However, I also understand that contaminating "signature space" of `IndexStyle` is not particularly a clean solution.

If we say that `IndexStyle` for `Broadcasted` has to be a proper trait, we need to note that `IndexStyle` must be properly defined whenever `axes` is customized since [`axes` is a public interface](https://docs.julialang.org/en/latest/manual/interfaces/#man-interfaces-broadcasting-1).

File: base/reduce.jl
Comment (tkf) [reply base/reduce.jl]: If we go with trait direction, one option is to use `IndexStyle`.  Something like this:

```julia
struct NonIndexable <: IndexStyle end
IndexStyle(::Type) = NonIndexable()
```

File: base/broadcast.jl
Comment (nalimilan) [reply base/broadcast.jl]: Yeah, but as long as you define `IndexStyle` it should be a proper trait, independent from how it's used internally. Otherwise, you'd better define a custom internal function.

Anyway, what's the problem with the current implementation in this PR? Just that's it's complicated?

Comment (tkf) [reply base/broadcast.jl]: OK. So, reading the documentation and the code more, it looks like `Base.axes(::Broadcasted{MyStyle})` is not an overloadable method?  From the documentation, `Base.axes(x)` is overloadable if `x` participate in the broadcasting.  This method is then called via `combine_axes` in `axes(::Broadcasted)`.  I was initially worried that `IndexStyle(::Type{<:Broadcasted})` and `Base.axes(::Broadcasted)` can become incompatible if downstream package authors are not careful but it looks like I don't need to worry about it.

File: base/broadcast.jl
Comment (tkf) [inline base/broadcast.jl]: It turned out my previous approach (combine `IndexStyle` of each argument) was wrong since it didn't work with, e.g., `broadcasted(+, zeros(5), zeros(1, 4))`.  Each argument can be `IndexLinear` even though the broadcasted indexing is not (that's the main job of broadcasted).

So now the new approach is to return an equivalent of `IndexStyle(bc.axes[1])` if `length(bc.axes) == 1` and then return `IndexCartesian()` otherwise.  As I need `Axes` type parameter now, this implementation requires the `Broadcasted` to be `instantiate`d first.  I guess this is OK since that's the usual broadcasting pipeline.

Is this implementation fine?  Is `IndexStyle(bc.axes[1]) == IndexLinear() && length(bc.axes) == 1` the only possible `IndexLinear()` case that is detectable in a type-stable manner?  That is to say, I'm ignoring the case like `broadcasted(+, zeros(1, 4), zeros(1, 4))` as I need to look at the value to check that it supports linear indexing.

File: base/broadcast.jl
Comment (mbauman) [inline base/broadcast.jl]:
```suggestion
Base.IndexStyle(::Type{<:Broadcasted{<:Any,Tuple{Any}}}) = IndexLinear()
```

It doesn't really matter how the particular axis is implemented — we can still just use one linear index to access into it.

File: base/broadcast.jl
Comment (mbauman) [reply base/broadcast.jl]: This doesn't really require the `Broadcasted` to be `instantiate`d to get an `IndexStyle` — it just means that if it's not instantiated that it falls back to a cartesian implementation. All `Broadcasted`s should support cartesian indexing; only one-dimensional ones will allow straight integers.

File: base/broadcast.jl
Comment (mbauman) [inline base/broadcast.jl]:
```suggestion
```

We cannot define `LinearIndices` for `Broadcasted`s since they don't support linear indexing. They're not `<: AbstractArray`, so they don't the fancy linear->cartesian methods that do this conversion.  Did you need this to make any of the reductions work?

File: base/reduce.jl
Comment (mbauman) [reply base/reduce.jl]: That's a clever idea.  I think the union here is sensible for now — that trait could possibly come later.

File: base/broadcast.jl
Comment (tkf) [reply base/broadcast.jl]: Yes, we need it for this `_mapreduce` to work:

https://github.com/JuliaLang/julia/blob/85f6405c8d81a2de2663034861e10a5668beddbd/base/reduce.jl#L305-L306

Why `Broadcasted`s "don't support linear indexing"?  I thought it's fine because we get to this `_mapreduce` implementation only if `IndexStyle(::Type{<:Broadcasted})` returns `IndexLinear()`.  I thought `IndexStyle(x) == IndexLinear()` implies that `LinearIndices(x)` works.  Or am I misunderstanding the relationship between `IndexLinear` and `LinearIndices`?

File: base/reduce.jl
Comment (tkf) [reply base/reduce.jl]: Differing the whole designing work for the trait makes sense.

File: base/broadcast.jl
Comment (mbauman) [reply base/broadcast.jl]: We've not very carefully laid out all the restrictions on what these functions mean for non-`AbstractArray` types, but my rule would be that:

```
A == A[LinearIndices(A)] == A[CartesianIndices(A)]
```

Defining this for all Braodcasteds would totally break that.  If we just need this for `IndexLinear` ones, then lets define it as:

```suggestion
Base.LinearIndices(bc::Broadcasted{<:Any,Tuple{Any}}) = axes(bc)[1]
```

This way it'd be either an outright error or indexing with the result will work.

File: base/broadcast.jl
Comment (tkf) [reply base/broadcast.jl]: Yes, that's much more accurate way to put it.  I wanted to say that "for the broadcasted to be dispatched to the optimized method", it has to be `instantiate`d first.

File: base/broadcast.jl
Comment (tkf) [reply base/broadcast.jl]: Ah, yes, thanks, this definition makes sense. It was wrong to define `LinearIndices` _for all_ `Broadcasted` instances.

**Comment by chethega** [comment]
Can we also speed up `all`, `any` and `count`?

These reductions currently don't use the mapreduce framework and therefore have to be done separately.

**Comment by tkf** [comment]
I think `all` and `any` use `mapreduce`:

https://github.com/JuliaLang/julia/blob/b1acb3c9c94214e211feddb799909760703f5b48/base/reducedim.jl#L664-L677

`count` doesn't, but it should be easy to add:

https://github.com/JuliaLang/julia/blob/111b385ea773d2a0bb034f1a13968d63b91f5c4e/base/reduce.jl#L749-L755

**Comment by tkf** [comment]
@mbauman I added some tests for this feature. This is all I wanted to do in this PR. Please review and merge or let me know if there are anything I need to fix.

**Comment by mbauman** [comment]
This patch looks good to me and seems worth doing.  I'd like to hold off just a bit to see what the triage reactions are to #19198 (and its implementation).

**Comment by StefanKarpinski** [comment]
Marked for triage so that it actually gets discussed.

**Comment by mbauman** [comment]
Yeah, I just put #31088 on the triage list earlier this afternoon — that's more the thing to talk about.

**Comment by mbauman** [comment]
Triage agrees this this is worth doing once we finalize #31088.

File: base/reduce.jl
Comment (mbauman) [reply base/reduce.jl]: @vtjnash mentioned there may be another way around this bootstrapping issue.

**Comment by tkf** [comment]
Cool. Thanks a lot for discussing it in triage.

**Comment by tkf** [comment]
As @StefanKarpinski mentioned in https://github.com/JuliaLang/julia/issues/19198#issuecomment-487989797, it makes sense to explore the design for the syntax to create `Broadcasted` #19198.  Meanwhile, would it be possible to merge this before the syntax is ready?  I think it makes sense because:

1. `Broadcasted` can already be constructed by users only using the public APIs documented.
2. Libraries such as [LazyArrays.jl](https://github.com/JuliaArrays/LazyArrays.jl) already have support for `Broadcasted`.  In particular, LazyArrays.jl has macro `@~` that is a super-set of the macro `@:` I proposed in #31020 to create a `Broadcasted`.
3. This specialization must happen in `Base` to avoid type piracy.  It's something impossible or very difficult to do outside `Base`.
4. If this PR is merged we can also start adding more specializations like handling multi-dimensional `Broadcasted`.  It would make `mapreduce` for `Broadcasted` applicable for many situations by the time the syntax or macro is included in the language.

**Comment by StefanKarpinski** [comment]
@mbauman, can you take on this one?

**Comment by tkf** [comment]
Resolved the merge conflict and also extended the definition to `reduce` with `dims`.

File: base/reduce.jl
Comment (nalimilan) [reply base/reduce.jl]: Given that the 1.4 feature freeze is approaching, I support merging the PR with the `Union` for now, and try to use a more general trait later.

FWIW, this PR is giving us incredibly good performance to be able to reuse the pairwise summation algorithm with weights at https://github.com/JuliaStats/StatsBase.jl/issues/518. Great work!

Comment (tkf) [reply base/reduce.jl]: Thanks a lot for testing this PR in a real-world scenario!

**Comment by oschulz** [comment]
This is awesome - we tried it in the context of JuliaStats/StatsBase.jl#518, in cases like that, this PR will help a lot. Do you guys think this can be merged in time for Julia 1.4?

**Comment by tkf** [comment]
friendly bump

@mbauman Can you review this?

**Comment by oschulz** [comment]
Just curious, any news on this? I saw some nice performance improvements with this.

**Comment by tkf** [comment]
I think there are two questions:

(Q1) Can this PR be merged without a short-hand syntax/macro to construct `Broadcasted` objects?

(Q2) Is my implementation OK?

While it is better to wait for @mbauman to look into it to answer Q2, I think other core devs can answer Q1 (or maybe it can be briefly discussed in triage?). Maybe @StefanKarpinski can help to clarify Q1?

I'm bringing this up as it seems that triage wanted to implement the syntax at the same time https://github.com/JuliaLang/julia/pull/31020#issuecomment-473033327 and if the answer to Q1 is no, we need to do some more discussion in #19198 while waiting for Q2. It also gives people some idea of how long they need to wait to get this.  If this requires a syntax, it's imaginable that this takes way longer than just adding an overload.

**Comment by PhilipVinc** [comment]
Out of curiosity: I just noticed that in 1.4-rc2
`Juno.@enter sum(Broadcast.broadcasted(*, rand(10,20), rand(10,20)), dims=1)` fails with a no method error.

Since this PR is relevant, does it fix this issue?

(it's a shame it was not included in v1.4)

**Comment by tkf** [comment]
I guess `sum` with `dims` is implemented only for arrays before this PR. So, it's relevant in the sense that this PR makes it work.

But I'm not sure about `Juno.@enter` part. If it's something debugger-specific, maybe report it to Juno or Debugger.jl?

**Comment by tkf** [comment]
Actually, no, this PR does not fix it (yet). It should be easy to add, though.

**Comment by PhilipVinc** [comment]
Sorry, my bad. The `Juno.@enter` part I left after copy pasting. Of course it's completely irrelevant to my question.

But I after looking through your this PR, I noticed you implement `Base.mapreducedim!` which is what sum with dims calls under the hood, so I thought it should work. But I'm probably wrong.

**Comment by tkf** [comment]
Yeah, we have all the internals we need. But I think we still need to define the dispatch.  For `sum` etc., this is just replacing `AbstractArray` with `AbstractArrayOrBroadcasted` in

https://github.com/JuliaLang/julia/blob/94b29d5e9849a915a99f05ca66450770c6ed292e/base/reducedim.jl#L648-L653

I'm suspecting there is a somewhat long list of functions receiving `dims`. So I think I'll do it just before or after this PR is merged (as otherwise, it can cause conflict with other PRs).

**Review by mbauman — APPROVED** [review]

**Comment by mbauman** [comment]
So to (finally) answer your two questions:

* Q1 (Can this PR be merged without a short-hand syntax/macro to construct Broadcasted objects?): Absofreakenlutely
* Q2 (Is my implementation OK?): Yes, looks great to me.

**Comment by tkf** [comment]
Great! Thanks a lot for reviewing this.

**Comment by oschulz** [comment]
Now it's approved - Is there a chance for this to make it into 1.5?

**Comment by KristofferC** [comment]
It would be good re-rebase this on master so that CI reruns on top of that (since it was a while CI ran).

**Comment by tkf** [comment]
I suppose merging `master` to this branch should be fine (as the commits are going to be squashed)? I can rebase, of course, if it's better.

File: base/broadcast.jl
Comment (martinholters) [inline base/broadcast.jl:230]: Am right that this usually won't return a `LinearIndices`? Should this rather be `LinearIndices(axes(bc))`?

Comment (mbauman) [reply base/broadcast.jl:230]: Yeah, that'd probably be better.  This sort of specialization should probably be on `eachindex(::IndexLinear, ...)` or `LinearIndices` itself if it's even needed at all.

**Pull request merged**
